src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,436 @@
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::sendRequested,
+            this,
+            &ControllerRenderingEngine::send);
+    connect(this,
+            &ControllerRenderingEngine::stopRequested,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_pThread->start(QThread::NormalPriority);
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.critical() << "The ControllerEngine is being deleted but hasn't been "
+                              "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!thread()->isFinished() && !thread()->isInterruptionRequested()) {
+        kLogger.critical() << "Render thread has or is about to terminate. Cannot "
+                              "start this render anymore.";
+        return;
+    }
+    QCoreApplication::postEvent(this, new QEvent(QEvent::UpdateRequest));
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pThread && m_pThread->isRunning();
+}
+
+void ControllerRenderingEngine::requestSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
+    VERIFY_OR_DEBUG_ASSERT(qmlEngine) {
+        kLogger.critical() << "No QML engine was passed!";
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(QThread::currentThread() != thread()) {
+        kLogger.warning() << "Unable to setup OpenGL rendering context from the same "
+                             "thread as the render object";
+        return;
+    }
+    emit setupRequested(qmlEngine);
+
+    const auto lock = lockMutex(&m_mutex);
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+    if (m_isValid) {
+        m_renderControl->prepareThread(m_pThread.get());
+    }
+}
+
+void ControllerRenderingEngine::requestSend(Controller* controller, const QByteArray& frame) {
+    emit sendRequested(controller, frame);
+}
+
+void ControllerRenderingEngine::setup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    QSurfaceFormat format;
+    format.setSamples(m_screenInfo.msaa);
+    format.setDepthBufferSize(16);
+    format.setStencilBufferSize(8);
+
+    const auto lock = lockMutex(&s_glMutex);
+
+    m_context = std::make_unique<QOpenGLContext>();
+    m_context->setFormat(format);
+    VERIFY_OR_DEBUG_ASSERT(m_context->create()) {
+        kLogger.warning() << "Unable to initialize controller screen rendering. Giving up";
+        m_waitCondition.wakeAll();
+        return;
+    }
+    connect(m_context.get(),
+            &QOpenGLContext::aboutToBeDestroyed,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_offscreenSurface = std::make_unique<QOffscreenSurface>();
+    m_offscreenSurface->setFormat(m_context->format());
+
+    VERIFY_OR_DEBUG_ASSERT(QMetaObject::invokeMethod(
+                                   qApp,
```

Comment:
This is defined in `<QCoreApplication>`, it's a Qt macro which expand to `QCoreApplication::instance()`

Reply:
Interesting. I'd prefer the explicit QCoreApplication::instance() tbh. Just because `qApp` is not commonly used in mixxx and its really difficult to look up where it comes up (if ur not looking at the code from an editor with intellisense)

Reply:
`qApp` is a [common macro](https://doc.qt.io/qt-6/qapplication.html#macros) and used at various place in Mixxx already. Is there a consensus on banning that macro on new code?

Reply:
okay, then disregard my comment. I've personally never encountered it so I got tripped up by it (and some googling didn't help). Feel free to keep it then.